fedimint/fedimint
# feat: clients use vetted gateway meta field in selecting gateway

Fedimint clients select gateways based on [`vetted_gateways` meta field](https://github.com/fedimint/fedimint/blob/master/docs/meta_fields/vetted_gateways.md)
as configured for a federation via some `meta_override_url`

**Testing:**
- Configure FM  with`meta_override_url`. For example in devimint, you can add the following to [federation.rs](https://github.com/fedimint/fedimint/blob/c5ce0e99cb6ee18e3e975578891e9b0f752b327e/devimint/src/federation.rs#L501) and then `just mprocs`

```rs
let mut meta = BTreeMap::from([
    ("federation_name".to_string(), "testfed".to_string()),
    (
        "meta_override_url".to_string(),
        serde_json::json!("https://getln.xyz/fedimint/meta").to_string(),
    ),
])
```
- Get the `federation-id` and one or more `gateway-id`s then update `vetted_gateways` served by meta override url
```
curl -X POST https://getln.xyz/fedimint/meta -H "Content-Type: application/json" -d '{
    "federation-id": {
        "vetted_gateways": ["gateway-id"]
    }
}'
```
This federation id and vetted gateways should show up in `curl -X POST https://getln.xyz/fedimint/meta` json

- Run `fedimint-cli list-gateways` and confirm the vetted gateways are visible to the client
- Remove one or all gateways from the vetted list
```
curl -X POST https://getln.xyz/fedimint/meta -H "Content-Type: application/json" -d '{
    "federation-id": {
        "vetted_gateways": []
    }
}'
```

Meta configs are **cached up to 10 minutes** by the client

closes #3420, closes #3649
